tests/e2e/__testers__.ts

```diff
@@ -4,63 +4,167 @@ import { values } from 'lodash'
 import stripAnsi from 'strip-ansi'
 import { ClientBase } from '@prisma-spectrum/reflector/dist-cjs/Client'
 import { Reflector } from '@prisma-spectrum/reflector'
+import { PrismaClient } from '@prisma/client'
+import { log } from 'floggy'
+import { PrismaClientConstructor } from '~/tests/e2e/helpers/getMysqlAdminPrismaClient'
+import { casesHandled } from '~/src/utils'
 
-export const testTemplate = (params: {
+export interface DBTestParams {
   templateName: PrismaTemplates.$Types.Template['_tag']
-  /** A string or regex pattern of the expected output (stdout) when the project dev mode is run.*/
   expectedDevOutput: RegExp | string
-}) => {
-  jest.setTimeout(50_000)
+  datasourceProvider: Reflector.Schema.DatasourceProviderNormalized
+  prismaConfig?: {
+    referentialIntegrity?: 'prisma' | 'foreignKeys'
+  }
+}
+
+async function dropDatabase(
+  prismaClient: PrismaClient,
+  databaseName: string,
+  datasourceProvider: Reflector.Schema.DatasourceProviderNormalized
+) {
+  switch (datasourceProvider) {
+    case 'postgres':
+      try {
+        await prismaClient.$executeRawUnsafe(`DROP DATABASE ${databaseName} WITH (FORCE);`)
+      } catch (error) {
+        const isDatabaseNotFoundErorr = error instanceof Error && error.message.match(/does not exist/)
+        if (!isDatabaseNotFoundErorr) throw error
+      }
+      return
+    case 'mysql':
+      try {
+        await prismaClient.$executeRawUnsafe(`DROP DATABASE IF EXISTS ${databaseName};`)
+      } catch (error) {
+        const isDatabaseNotFoundErorr = error instanceof Error && error.message.match(/database not found/)
+        if (!isDatabaseNotFoundErorr) throw error
+      }
+      return
+    case 'cockroachdb':
+    case 'mongodb':
+    case 'sqlserver':
+    case 'sqlite':
+      throw new Error(`Not implemented yet.`)
+    default:
+      casesHandled(datasourceProvider)
+  }
+}
+
+async function createDatabase(
+  prismaClient: PrismaClient,
+  databaseName: string,
+  datasourceProvider: Reflector.Schema.DatasourceProviderNormalized
+) {
+  switch (datasourceProvider) {
+    case 'postgres':
+      try {
+        await prismaClient.$executeRawUnsafe(`create database ${databaseName}`)
+      } catch (e) {
+        log.info(`Error initialising DB ${databaseName}`)
+      }
+      return
+    case 'mysql':
+      return prismaClient.$executeRawUnsafe(`CREATE DATABASE IF NOT EXISTS ${databaseName}`)
+    default:
+      throw new Error(`Case not handled for ${datasourceProvider}`)
+  }
+}
+
+export function getConnectionString(
+  dataSourceProvider: Reflector.Schema.DatasourceProviderNormalized
+): string {
+  switch (dataSourceProvider) {
+    case 'postgres':
+      return 'postgres://prisma:prisma@localhost:5401'
+    case 'mysql':
+      return 'mysql://prisma:prisma@localhost:33577'
+    default:
+      throw new Error(`Case not handled for ${dataSourceProvider}`)
```

Comment:
This is doing something at runtime we can catch statically. Look at other switch pattern.

Reply:
Not sure what you mean? The data source provider is being passed down from the test itself, meaning it can only fail on runtime.

Could you share an example on how you think this should look like?

Reply:
Right now this default does two things implicitly:

1. throw error for datasources we know we do not support yet
2. throw error for datasources that have been added but we did not think about yet, because they have been newly added to the union.

If we're ok with this then there's nothing to do. If however we want to be more explicit than we can do here what we did with the other switch statement.